update readme (#4)
DevExpress-Examples/XAF_How-to-get-role-code-from-the-UI: Readme.md

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T868197)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->

> # How to generate database updater code for security roles created via the application UI in a development environment

Reviewer: ```suggestion
# How to generate database updater code for security roles created in the application UI in a development environment

```

![image](https://user-images.githubusercontent.com/14300209/77691659-62c48a00-6fb6-11ea-9d52-d273a30c137d.png)

>If this solution does not work for you, you can dump required data records from the development database to an XML file and then load this XML file from an application using the production database or transfer data using the built-in RDBMS capabilities. For more information, please review [Security - Best Practices for Export/Import Role Permissions at runtime (without releasing a new application version to clients)](https://supportcenter.devexpress.com/ticket/details/t951640/security-best-practices-for-export-import-role-permissions-at-runtime-without-releasing).

> Note: You can find the solution for Full Framework, WebForms, VB.NET in the [20.1.3 branch](https://github.com/DevExpress-Examples/XAF_How-to-get-role-code-from-the-UI/tree/20.1.3+).

## Implementation Details

> **Step 1.** In the Solution Explorer, include [RoleGenerator.csproj](CS/EFCore/GenerateRoleEF/RoleGenerator/RoleGenerator.csproj) into your XAF solution and then reference this *RoleGenerator* project from the *YourSolutionName.Module* one.

Reviewer: ```suggestion
In the Solution Explorer, include [RoleGenerator.csproj](CS/EFCore/GenerateRoleEF/RoleGenerator/RoleGenerator.csproj) into your XAF solution.

In the **YourSolutionName.Module** project, add a reference to the **RoleGenerator** project.

```

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T868197)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->

# XAF - Generate Database Updater Code for Security Roles Created in Application UI in Development Environment

> XAF developers often create initial security roles in the administrative UI at runtime in non-production databases of test environments. The visual approach is often faster than writing code manually, especially for [complex permissions with criteria](https://docs.devexpress.com/eXpressAppFramework/113366/data-security-and-safety/security-system#architecture). If you create initial roles at runtime in test databases, at some point, you may need to transfer the data to production databases on customer sites. 

Reviewer 2: The first sentence is not especially long, but it contains a lot of information to process: security roles, runtime admin UI, non-production database, test environment. I wanted to split it up in a few sentences, so that it's easier to follow. 

```suggestion
An XAF test environment usually uses a non-production database. Developers often populate such databases with initial security roles. They use runtime administrative UI for that purpose, since the visual approach is often faster than writing code, especially for [complex permissions with criteria](https://docs.devexpress.com/eXpressAppFramework/113366/data-security-and-safety/security-system#architecture). At some point, developers may need to transfer role data to production databases on customer sites. 
```

> [ModuleUpdater](https://docs.devexpress.com/eXpressAppFramework/DevExpress.ExpressApp.Updating.ModuleUpdater) API and [DBUpdater](https://docs.devexpress.com/eXpressAppFramework/113239/deployment/deployment-tutorial/application-update#update-a-database-dbupdater-tool) are standard means to seed initial data in databases with XAF.

Reviewer 2: This sentences sounds too "detached" from the main topic. We were just describing a possible task, we need to transition to a solution. And this sentence is just purely theoretical - just a standalone sentence that doesn't seem to have any connection to the previous paragraph. 

As a reader, you can figure out the intent. But it's better to be more direct in sending that message. 

```suggestion
This example relies on standard XAF mechanisms that help you seed initial data in databases: [ModuleUpdater](https://docs.devexpress.com/eXpressAppFramework/DevExpress.ExpressApp.Updating.ModuleUpdater) API and [DBUpdater](https://docs.devexpress.com/eXpressAppFramework/113239/deployment/deployment-tutorial/application-update#update-a-database-dbupdater-tool)  tool.
```

> This example demonstrates how to automatically create `ModuleUpdater` code for the required roles created by XAF developers at runtime in test databases. You can easily copy and paste this code into your `ModuleUpdater` descendant and use the standard `DBUpdater` tool to seed data in production databases. We intentionally skip user creation code, because user names are often unknown at early stages and it is much easier to create and link users to predefined roles later in production environment.

Reviewer 2: I'm already using "This example" as the beginning of the previous paragraph. And I also wanted to be a little more detailed about the overall mechanism. 

I'm also separating information about "user creation code" into a separate paragraph. Seems like that thought deserves its own space. 

```suggestion
The transfer mechanism suggested in this solution works in the following manner. You embed controllers from this example into your test application. These embedded controllers can analyze database content and generate `ModuleUpdater` code for required roles. You can then copy and paste this code into your production project's `ModuleUpdater` descendant. Use the standard `DBUpdater` tool to seed data in the database. 

The example intentionally skips user creation code. User names are often unknown at early stages and it's much easier to create and link users to predefined roles later in production environment.
```

> ![image](https://user-images.githubusercontent.com/14300209/77691659-62c48a00-6fb6-11ea-9d52-d273a30c137d.png)

Reviewer 2: The image has some strange black "waves" at the bottom and on the right.

Reviewer: Fixed.

> If this solution does not work for you, you can dump the required data records from the development database to an XML file and then load this XML file from an application using the production database or transfer the data using the built-in RDBMS capabilities. For more information, see the following Support Center ticket: [Security - Best Practices for Export/Import Role Permissions at runtime (without releasing a new application version to clients)](https://supportcenter.devexpress.com/ticket/details/t951640/security-best-practices-for-export-import-role-permissions-at-runtime-without-releasing). Please note, that this solution is applicable only if you use XPO.

Reviewer 2: I removed a few unnecessary articles (mainly with plural or uncountable nouns). And added a list - seems like it's a more readable version of that long sentence. 

```suggestion
If this solution does not work for you, you can use the following alternative methods: 

- Save required data records from the development database to an XML file and then load this XML file in an application that uses the production database.
- Transfer data using built-in RDBMS capabilities. 

For more information, see the following Support Center ticket: [Security - Best Practices for Export/Import Role Permissions at runtime (without releasing a new application version to clients)](https://supportcenter.devexpress.com/ticket/details/t951640/security-best-practices-for-export-import-role-permissions-at-runtime-without-releasing). Please note that this solution is applicable only if you use XPO.
```

> **Note**  

> > You can find the solution for .NET Framework, WebForms, and VB.NET in the [20.1.3 branch](https://github.com/DevExpress-Examples/XAF_How-to-get-role-code-from-the-UI/tree/20.1.3+).

Reviewer 2: ```suggestion
> You can find a solution for .NET Framework, Web Forms, and VB.NET in branch [20.1.3](https://github.com/DevExpress-Examples/XAF_How-to-get-role-code-from-the-UI/tree/20.1.3+).

```

	- Include the generated *Updater.xx* file into the *YourSolutionName.Module/DatabaseUpdate* folder and modify the [YourSolutionName/Module.cs](CS/EFCore/GenerateRoleEF/GenerateRoleEF.Module/Module.cs) file to use this new `RoleUpdater` class as follows:
 
	``` csharp
	// C#
	using System;
	using DevExpress.ExpressApp;
	using System.Collections.Generic;
	using DevExpress.ExpressApp.Updating;
	
	namespace YourSolutionName.Module {
	  public sealed partial class YourSolutionNameModule : ModuleBase {
	    public override IEnumerable<ModuleUpdater> GetModuleUpdaters(IObjectSpace objectSpace, Version versionFromDB) {
	      ModuleUpdater updater = new DatabaseUpdate.Updater(objectSpace, versionFromDB);
	        ModuleUpdater roleUpdater = new RoleUpdater(objectSpace, versionFromDB);
	    	return new ModuleUpdater[] { updater, roleUpdater };
	//...
	```

	> **Note**  
 	> In the ASP.NET Core Blazor application, the file is saved to the Documents folder by default. You can change this behavior by overriding the `RoleGeneratorControllerBlazor.SaveFile` method.

### Customization for Custom Security Roles

> If you have a custom security role class (for example, `ExtendedSecurityRole` implemented in the following examples: [Implement a Custom Security System User Based on an Existing Business Class](https://docs.devexpress.com/eXpressAppFramework/113452/task-based-help/security/how-to-implement-a-custom-security-system-user-based-on-an-existing-business-class), [Implement Custom Security Objects (Users, Roles, Operation Permissions)](https://docs.devexpress.com/eXpressAppFramework/113384/task-based-help/security/how-to-implement-custom-security-objects-users-roles-operation-permissions)) and this class has custom properties, you need to include these properties in the generated code. To do this, handle the `RoleGenerator.CustomizeCodeLines` event in the `RoleGeneratorController` class added at Step 2:

Reviewer 2: Again, just breaking it down into shorter, more manageable sentences. 

```suggestion
You may use a custom security role class. For example, `ExtendedSecurityRole` implementations are available in the following examples: [Implement a Custom Security System User Based on an Existing Business Class](https://docs.devexpress.com/eXpressAppFramework/113452/task-based-help/security/how-to-implement-a-custom-security-system-user-based-on-an-existing-business-class), [Implement Custom Security Objects (Users, Roles, Operation Permissions)](https://docs.devexpress.com/eXpressAppFramework/113384/task-based-help/security/how-to-implement-custom-security-objects-users-roles-operation-permissions). If a security role has custom properties, you need to include these properties in the generated code. To do this, handle the `RoleGenerator.CustomizeCodeLines` event in the `RoleGeneratorController` class added in Step 2:

```

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T868197)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->

# XAF - Generate Database Updater Code for Security Roles Created in Application UI in Development Environment

> An XAF test environment usually uses a non-production database. Developers often populate such databases with initial security roles. They use runtime administrative UI for that purpose, since the visual approach is often faster than writing code, especially for [complex permissions with criteria](https://docs.devexpress.com/eXpressAppFramework/113366/data-security-and-safety/security-system#architecture). At some point, developers may need to transfer role data to production databases on customer sites. 

Reviewer 3: ```suggestion
An XAF test environment usually uses a non-production database. Developers often populate such databases with initial security roles. They use a runtime administrative UI to do this since the visual approach is often faster than writing code, especially for [complex permissions with criteria](https://docs.devexpress.com/eXpressAppFramework/113366/data-security-and-safety/security-system#architecture). At some point, developers may need to transfer role data to production databases on customer sites. 

```

This example relies on standard XAF mechanisms that help you seed initial data in databases: [ModuleUpdater](https://docs.devexpress.com/eXpressAppFramework/DevExpress.ExpressApp.Updating.ModuleUpdater) API and [DBUpdater](https://docs.devexpress.com/eXpressAppFramework/113239/deployment/deployment-tutorial/application-update#update-a-database-dbupdater-tool)  tool.

> The transfer mechanism suggested in this solution works in the following manner. You embed controllers from this example into your test application. These embedded controllers can analyze database content and generate `ModuleUpdater` code for required roles. You can then copy and paste this code into your production project's `ModuleUpdater` descendant. Use the standard `DBUpdater` tool to seed data in the database. 

Reviewer 3: ```suggestion
The transfer mechanism suggested in this solution works as follows: you embed controllers from this example into your test application, and these embedded controllers can analyze database content and generate `ModuleUpdater` code for required roles. You can then copy and paste this code into your production project's `ModuleUpdater` descendant. Use the standard `DBUpdater` tool to seed data in the database. 

```

> The example intentionally skips user creation code. User names are often unknown at early stages and it's much easier to create and link users to predefined roles later in production environment.

Reviewer 3: ```suggestion
The example intentionally skips user creation code. User names are often unknown at early stages which simplifies creating and linking users to predefined roles later in a production environment.

```

![image](/Images/xaf-generate-database-updater-code-from-ui-devexpress.png)

> If this solution does not work for you, you can use the following alternative methods: 

Reviewer 3: ```suggestion
If this solution is not suitable, you can use one of the following alternatives: 

```

> - Save required data records from the development database to an XML file and then load this XML file in an application that uses the production database.

Reviewer 3: ```suggestion
- Save the data records from the development database in an XML file and then load this XML file in an application that uses the production database.

```

- Transfer data using built-in RDBMS capabilities. 

For more information, see the following Support Center ticket: [Security - Best Practices for Export/Import Role Permissions at runtime (without releasing a new application version to clients)](https://supportcenter.devexpress.com/ticket/details/t951640/security-best-practices-for-export-import-role-permissions-at-runtime-without-releasing). Please note that this solution is applicable only if you use XPO.

> **Note**  
> You can find a solution for .NET Framework, Web Forms, and VB.NET in branch [20.1.3](https://github.com/DevExpress-Examples/XAF_How-to-get-role-code-from-the-UI/tree/20.1.3+).

## Implementation Details

> 1. In the Solution Explorer, include [RoleGenerator.csproj](CS/EFCore/GenerateRoleEF/RoleGenerator/RoleGenerator.csproj) into your XAF solution.

Reviewer 3: ```suggestion
1. In the Solution Explorer, include [RoleGenerator.csproj](CS/EFCore/GenerateRoleEF/RoleGenerator/RoleGenerator.csproj) in your XAF solution.

```

	- Include the generated *Updater.xx* file into the *YourSolutionName.Module/DatabaseUpdate* folder and modify the [YourSolutionName/Module.cs](CS/EFCore/GenerateRoleEF/GenerateRoleEF.Module/Module.cs) file to use this new `RoleUpdater` class as follows:
 
	``` csharp
	// C#
	using System;
	using DevExpress.ExpressApp;
	using System.Collections.Generic;
	using DevExpress.ExpressApp.Updating;
	
	namespace YourSolutionName.Module {
	  public sealed partial class YourSolutionNameModule : ModuleBase {
	    public override IEnumerable<ModuleUpdater> GetModuleUpdaters(IObjectSpace objectSpace, Version versionFromDB) {
	      ModuleUpdater updater = new DatabaseUpdate.Updater(objectSpace, versionFromDB);
	        ModuleUpdater roleUpdater = new RoleUpdater(objectSpace, versionFromDB);
	    	return new ModuleUpdater[] { updater, roleUpdater };
	//...
	```

	> **Note**  
 	> In the ASP.NET Core Blazor application, the file is saved to the Documents folder by default. You can change this behavior by overriding the `RoleGeneratorControllerBlazor.SaveFile` method.

### Customization for Custom Security Roles

> You may use a custom security role class. For example, `ExtendedSecurityRole` implementations are available in the following examples: [Implement a Custom Security System User Based on an Existing Business Class](https://docs.devexpress.com/eXpressAppFramework/113452/task-based-help/security/how-to-implement-a-custom-security-system-user-based-on-an-existing-business-class), [Implement Custom Security Objects (Users, Roles, Operation Permissions)](https://docs.devexpress.com/eXpressAppFramework/113384/task-based-help/security/how-to-implement-custom-security-objects-users-roles-operation-permissions). If a security role has custom properties, you need to include these properties in the generated code. To do this, handle the `RoleGenerator.CustomizeCodeLines` event in the `RoleGeneratorController` class added in Step 2:

Reviewer 3: ```suggestion
You can use a custom security role class. For example, `ExtendedSecurityRole` implementations are available in the following examples: [Implement a Custom Security System User Based on an Existing Business Class](https://docs.devexpress.com/eXpressAppFramework/113452/task-based-help/security/how-to-implement-a-custom-security-system-user-based-on-an-existing-business-class), [Implement Custom Security Objects (Users, Roles, Operation Permissions)](https://docs.devexpress.com/eXpressAppFramework/113384/task-based-help/security/how-to-implement-custom-security-objects-users-roles-operation-permissions). If a security role has custom properties, you need to include these properties in the generated code. To do this, handle the `RoleGenerator.CustomizeCodeLines` event in the `RoleGeneratorController` class added in Step 2:

```

}

```
> [!WARNING]

> > We created this example for demonstration purposes and it is not intended to address all possible usage scenarios with it.

Reviewer 2: ```suggestion
> We created this example for demonstration purposes and it is not intended to address all possible usage scenarios.

```

> > If this example does not have certain functionality or you want to change its behavior, you can extend this example as needed. Please note that this can be a complex task that requires good knowledge of XAF: [UI Customization Categories by Skill Level](https://www.devexpress.com/products/net/application_framework/xaf-considerations-for-newcomers.xml#ui-customization-categories). You will likely need to research how our components work under the hood. Refer to the following help topic for more information: [Debug DevExpress .NET Source Code with PDB Symbols](https://docs.devexpress.com/GeneralInformation/403656/support-debug-troubleshooting/debug-controls-with-debug-symbols).

Reviewer 2: Not a big fan of "under the hood" - kind of slang language coming out from nowhere. The language wasn't that conversational all this time, so I'd keep it formal.
 
```suggestion
> You can extend this example or change its behavior as needed. Please note that this can be a complex task that requires good knowledge of XAF: [UI Customization Categories by Skill Level](https://www.devexpress.com/products/net/application_framework/xaf-considerations-for-newcomers.xml#ui-customization-categories). You will likely need to research the internal architecture of DevExpress components. Refer to the following help topic for more information: [Debug DevExpress .NET Source Code with PDB Symbols](https://docs.devexpress.com/GeneralInformation/403656/support-debug-troubleshooting/debug-controls-with-debug-symbols).
```

> > We are unable to help with such tasks as custom programming is outside our Support Service scope: [Technical Support Scope](https://www.devexpress.com/products/net/application_framework/xaf-considerations-for-newcomers.xml#support).

Reviewer 2: ```suggestion
> We are unable to help with such tasks. Custom programming is outside our Support Service scope: [Technical Support Scope](https://www.devexpress.com/products/net/application_framework/xaf-considerations-for-newcomers.xml#support).

```

}

```
> [!WARNING]
> We created this example for demonstration purposes and it is not intended to address all possible usage scenarios.

> > You can extend this example or change its behavior as needed. Please note that this can be a complex task that requires good knowledge of XAF: [UI Customization Categories by Skill Level](https://www.devexpress.com/products/net/application_framework/xaf-considerations-for-newcomers.xml#ui-customization-categories). You will likely need to research the internal architecture of DevExpress components. Refer to the following help topic for more information: [Debug DevExpress .NET Source Code with PDB Symbols](https://docs.devexpress.com/GeneralInformation/403656/support-debug-troubleshooting/debug-controls-with-debug-symbols).

Reviewer 3: ```suggestion
> You can extend this example or change its behavior as needed. This can be a complex task that requires good knowledge of XAF: [UI Customization Categories by Skill Level](https://www.devexpress.com/products/net/application_framework/xaf-considerations-for-newcomers.xml#ui-customization-categories), and you might also need to research the internal architecture of DevExpress components. Refer to the following help topic for more information: [Debug DevExpress .NET Source Code with PDB Symbols](https://docs.devexpress.com/GeneralInformation/403656/support-debug-troubleshooting/debug-controls-with-debug-symbols).

```
